File: backend/app/api/v1/store/meetings.py
"""Meetings REST API (replaces frontend direct Supabase access).

Follows the deals.py reference: scoped reads via app/db/scope, single-row
fetches via scoped_*_or_404. No RLS — every query is org/meeting scoped here.
"""

from __future__ import annotations

from datetime import datetime

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy import select
from sqlalchemy.orm import Session

from app.api.v1.store._common import (
    get_db,
    get_principal,
    scoped_deal_or_404,
    scoped_meeting_or_404,
)
from app.db.models import Deal, Meeting
from app.db.scope import Principal, org_scoped
from app.schemas.common import BaseSchema

router = APIRouter()


# ---- schemas --------------------------------------------------------------
class MeetingResponse(BaseSchema):
    id: str
    org_id: str
    deal_id: str | None = None
    title: str
    meeting_date: datetime | None = None
    duration_seconds: int | None = None
    source: str
    source_url: str | None = None
    file_key: str | None = None
    status: str
    error_message: str | None = None
    bot_enabled: bool
    external_event_id: str | None = None
    external_provider: str | None = None
    created_by: str
    created_at: datetime
    updated_at: datetime


class MeetingUpdate(BaseSchema):
    title: str | None = None
    meeting_date: datetime | None = None
    bot_enabled: bool | None = None
    deal_id: str | None = None


class MeetingCreate(BaseSchema):
    title: str
    source: str = "upload"
    file_key: str | None = None
    source_url: str | None = None
    meeting_date: datetime | None = None
    duration_seconds: int | None = None
    bot_enabled: bool = True


class DealRef(BaseSchema):
    id: str
    name: str


class CalendarMeetingResponse(BaseSchema):
    id: str
    org_id: str
    deal_id: str | None = None
    title: str
    meeting_date: datetime | None = None
    duration_seconds: int | None = None
    source: str
    source_url: str | None = None
    file_key: str | None = None
    status: str
    error_message: str | None = None
    bot_enabled: bool
    external_event_id: str | None = None
    external_provider: str | None = None
    created_by: str
    created_at: datetime
    updated_at: datetime
    deal: DealRef | None = None


# ---- endpoints ------------------------------------------------------------
@router.get("/deals/{deal_id}/meetings", response_model=list[MeetingResponse])
def list_deal_meetings(
    deal_id: str,
    session: Session = Depends(get_db),
    principal: Principal = Depends(get_principal),
) -> list[MeetingResponse]:
    scoped_deal_or_404(session, principal, deal_id)
    rows = session.scalars(
        select(Meeting)
        .where(Meeting.deal_id == deal_id)
        .order_by(Meeting.created_at.desc())
    ).all()
    return [MeetingResponse.model_validate(m) for m in rows]


@router.post(
    "/deals/{deal_id}/meetings",
    response_model=MeetingResponse,
    status_code=status.HTTP_201_CREATED,
)
def create_meeting(
    deal_id: str,
    payload: MeetingCreate,
    session: Session = Depends(get_db),
    principal: Principal = Depends(get_principal),
) -> MeetingResponse:
    """Create a meeting under a deal (used by the upload-confirm flow)."""
    deal = scoped_deal_or_404(session, principal, deal_id)
    meeting = Meeting(
        org_id=deal.org_id,
        deal_id=deal.id,
        title=payload.title,
        source=payload.source,
        file_key=payload.file_key,
        source_url=payload.source_url,
        meeting_date=payload.meeting_date.isoformat() if payload.meeting_date else None,
        duration_seconds=payload.duration_seconds,
        bot_enabled=payload.bot_enabled,
        status="uploading",
        created_by=principal.user_id,
    )
    session.add(meeting)
    session.flush()
    return MeetingResponse.model_validate(meeting)


@router.get("/meetings/{meeting_id}", response_model=MeetingResponse)
def get_meeting(
    meeting_id: str,
    session: Session = Depends(get_db),
    principal: Principal = Depends(get_principal),
) -> MeetingResponse:
    return MeetingResponse.model_validate(
        scoped_meeting_or_404(session, principal, meeting_id)
    )


@router.patch("/meetings/{meeting_id}", response_model=MeetingResponse)
def update_meeting(
    meeting_id: str,
    payload: MeetingUpdate,
    session: Session = Depends(get_db),
    principal: Principal = Depends(get_principal),
) -> MeetingResponse:
    meeting = scoped_meeting_or_404(session, principal, meeting_id)
    data = payload.model_dump(exclude_unset=True)
    # A client-supplied deal_id must be a deal in the SAME org — never let a
    # meeting be reassigned to another tenant's deal (IDOR).
    if "deal_id" in data and data["deal_id"] is not None:
        target = scoped_deal_or_404(session, principal, data["deal_id"])
        if target.org_id != meeting.org_id:
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Cross-org deal")
    for field, value in data.items():
        if field == "meeting_date" and value is not None:
            value = value.isoformat()
        setattr(meeting, field, value)
    session.flush()
    return MeetingResponse.model_validate(meeting)


@router.get("/calendar/meetings", response_model=list[CalendarMeetingResponse])
def calendar_meetings(
    session: Session = Depends(get_db),
    principal: Principal = Depends(get_principal),
) -> list[CalendarMeetingResponse]:
    stmt = (
        org_scoped(select(Meeting, Deal), Meeting, principal)
        .outerjoin(Deal, Deal.id == Meeting.deal_id)
        .order_by(Meeting.meeting_date)
    )
    rows = session.execute(stmt).all()
    out: list[CalendarMeetingResponse] = []
    for meeting, deal in rows:
        resp = CalendarMeetingResponse.model_validate(meeting)
        resp.deal = DealRef(id=deal.id, name=deal.name) if deal is not None else None
        out.append(resp)
    return out


@router.get(
    "/dashboard/upcoming-unassigned", response_model=list[MeetingResponse]
)
def upcoming_unassigned(
    session: Session = Depends(get_db),
    principal: Principal = Depends(get_principal),
) -> list[MeetingResponse]:
    stmt = (
        org_scoped(select(Meeting), Meeting, principal)
        .where(Meeting.deal_id.is_(None))
        .where(Meeting.external_provider.is_not(None))
        .order_by(Meeting.meeting_date)
    )
    rows = session.scalars(stmt).all()
    return [MeetingResponse.model_validate(m) for m in rows]

File: worker/internal/store/meetings.go
package store

import (
	"context"
	"database/sql"
	"errors"
	"strings"

	"github.com/Yrock1x/AI-Notetaker-Finance/worker/internal/model"
	"github.com/Yrock1x/AI-Notetaker-Finance/worker/internal/util"
)

// ErrCrossOrg is returned when a meeting would be reassigned to a deal in a
// different org than the meeting itself; the handler maps it to 400 (ports the
// "Cross-org deal" HTTPException in update_meeting). It is meetings-specific, so
// it is NOT in the shared storeError() switch — handlers check it explicitly.
var ErrCrossOrg = errors.New("cross-org deal")

const meetingCols = "id, org_id, deal_id, title, meeting_date, duration_seconds, source, source_url, file_key, status, error_message, bot_enabled, external_event_id, external_provider, created_by, created_at, updated_at"

func scanMeeting(row interface{ Scan(...any) error }) (*model.Meeting, error) {
	var m model.Meeting
	err := row.Scan(&m.ID, &m.OrgID, &m.DealID, &m.Title, &m.MeetingDate,
		&m.DurationSeconds, &m.Source, &m.SourceURL, &m.FileKey, &m.Status,
		&m.ErrorMessage, &m.BotEnabled, &m.ExternalEventID, &m.ExternalProvider,
		&m.CreatedBy, &m.CreatedAt, &m.UpdatedAt)
	return &m, err
}

// ScopedMeeting returns a meeting owned by one of the principal's orgs, else
// ErrNotFound (ports scoped_meeting_or_404). A foreign/missing row is
// indistinguishable — both 404.
func ScopedMeeting(ctx context.Context, conn *sql.DB, p *Principal, meetingID string) (*model.Meeting, error) {
	pred, args := p.OrgFilter("org_id")
	q := "SELECT " + meetingCols + " FROM meetings WHERE id = ? AND " + pred
	m, err := scanMeeting(conn.QueryRowContext(ctx, q, append([]any{meetingID}, args...)...))
	if errors.Is(err, sql.ErrNoRows) {
		return nil, ErrNotFound
	}
	return m, err
}

// ListDealMeetings returns every meeting under a deal, newest-first (ports
// list_deal_meetings). The caller must have already ScopedDeal'd dealID; the
// query is additionally org-scoped here as defence-in-depth so a meeting can
// never leak across tenants even if the deal guard is skipped.
func ListDealMeetings(ctx context.Context, conn *sql.DB, p *Principal, dealID string) ([]model.Meeting, error) {
	pred, args := p.OrgFilter("org_id")
	q := "SELECT " + meetingCols + " FROM meetings WHERE deal_id = ? AND " + pred +
		" ORDER BY created_at DESC"
	rows, err := conn.QueryContext(ctx, q, append([]any{dealID}, args...)...)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	items := make([]model.Meeting, 0)
	for rows.Next() {
		m, err := scanMeeting(rows)
		if err != nil {
			return nil, err
		}
		items = append(items, *m)
	}
	return items, rows.Err()
}

// MeetingCreate is the create payload (ports MeetingCreate). Defaults
// (source="upload", status="uploading", bot_enabled=true) are applied here.
type MeetingCreate struct {
	Title           string
	Source          string
	FileKey         *string
	SourceURL       *string
	MeetingDate     *string
	DurationSeconds *int64
	BotEnabled      bool
}

// CreateMeeting inserts a meeting under a scoped deal (ports create_meeting).
// org_id is inherited from the deal — never from the client — so a meeting can
// only ever be created in the deal's own tenant.
func CreateMeeting(ctx context.Context, conn *sql.DB, p *Principal, deal *model.Deal, in MeetingCreate) (*model.Meeting, error) {
	now, id := util.NowISO(), util.NewUUID()
	source := in.Source
	if source == "" {
		source = "upload"
	}
	if _, err := conn.ExecContext(ctx,
		"INSERT INTO meetings("+meetingCols+") VALUES (?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)",
		id, deal.OrgID, deal.ID, in.Title, in.MeetingDate, in.DurationSeconds, source,
		in.SourceURL, in.FileKey, "uploading", nil, in.BotEnabled, nil, nil,
		p.UserID, now, now); err != nil {
		return nil, err
	}
	return ScopedMeeting(ctx, conn, p, id)
}

// MeetingUpdate carries the patchable fields. A nil pointer means "field absent
// from the request body" (the exclude_unset semantics); a present-but-null
// meeting_date/deal_id is expressed via the *Set flags so we can write SQL NULL.
type MeetingUpdate struct {
	Title          *string
	BotEnabled     *bool
	MeetingDate    *string
	MeetingDateSet bool // meeting_date was present in the body (value may be null)
	DealID         *string
	DealIDSet      bool // deal_id was present in the body (value may be null)
}

// UpdateMeeting patches the supplied fields on a scoped meeting (ports
// update_meeting). A client-supplied deal_id must resolve to a deal in the SAME
// org as the meeting, else ErrCrossOrg (400) — never let a meeting be reassigned
// into another tenant's deal (IDOR). Returns ErrNotFound if the target deal is
// missing/foreign.
func UpdateMeeting(ctx context.Context, conn *sql.DB, p *Principal, meetingID string, u MeetingUpdate) (*model.Meeting, error) {
	meeting, err := ScopedMeeting(ctx, conn, p, meetingID)
	if err != nil {
		return nil, err
	}

	// Reassigning to a deal: it must be visible to the principal AND share the
	// meeting's org.
	if u.DealIDSet && u.DealID != nil {
		target, err := ScopedDeal(ctx, conn, p, *u.DealID)
		if err != nil {
			return nil, err
		}
		if target.OrgID != meeting.OrgID {
			return nil, ErrCrossOrg
		}
	}

	var sets []string
	var args []any
	add := func(col string, v any) { sets = append(sets, col+" = ?"); args = append(args, v) }
	if u.Title != nil {
		add("title", *u.Title)
	}
	if u.BotEnabled != nil {
		add("bot_enabled", *u.BotEnabled)
	}
	if u.MeetingDateSet {
		add("meeting_date", u.MeetingDate) // *string -> NULL when nil
	}
	if u.DealIDSet {
		add("deal_id", u.DealID) // *string -> NULL when nil
	}
	if len(sets) > 0 {
		add("updated_at", util.NowISO())
		args = append(args, meetingID)
		if _, err := conn.ExecContext(ctx,
			"UPDATE meetings SET "+strings.Join(sets, ", ")+" WHERE id = ?", args...); err != nil {
			return nil, err
		}
	}
	return ScopedMeeting(ctx, conn, p, meetingID)
}
